src/main/java/ch/njol/skript/expressions/ExprDirection.java

```diff
@@ -36,177 +28,185 @@
+				} else if (isHorizontal) {
+					return new Direction[] {new Direction(new double[] {Math.cos(yaw) * meterLength, 0, Math.sin(yaw) * meterLength})};
+				} else if (isFacing) {
+					double pitch = Direction.pitchToRadians(relativeEntityLoc.getPitch());
 					assert pitch >= -Math.PI / 2 && pitch <= Math.PI / 2;
 					if (pitch > Math.PI / 4)
-						return new Direction[] {new Direction(new double[] {0, ln, 0})};
+						return new Direction[]{new Direction(new double[]{0, meterLength, 0})};
 					if (pitch < -Math.PI / 4)
-						return new Direction[] {new Direction(new double[] {0, -ln, 0})};
-				}
-				double yaw = Direction.yawToRadians(l.getYaw());
-				if (horizontal && !facing) {
-					return new Direction[] {new Direction(new double[] {Math.cos(yaw) * ln, 0, Math.sin(yaw) * ln})};
+						return new Direction[]{new Direction(new double[]{0, -meterLength, 0})};
 				}
-				yaw = Math2.mod(yaw, 2 * Math.PI);
-				if (yaw >= Math.PI / 4 && yaw < 3 * Math.PI / 4)
-					return new Direction[] {new Direction(new double[] {0, 0, ln})};
-				if (yaw >= 3 * Math.PI / 4 && yaw < 5 * Math.PI / 4)
-					return new Direction[] {new Direction(new double[] {-ln, 0, 0})};
-				if (yaw >= 5 * Math.PI / 4 && yaw < 7 * Math.PI / 4)
-					return new Direction[] {new Direction(new double[] {0, 0, -ln})};
-				assert yaw >= 0 && yaw < Math.PI / 4 || yaw >= 7 * Math.PI / 4 && yaw < 2 * Math.PI;
-				return new Direction[] {new Direction(new double[] {ln, 0, 0})};
+				return new Direction[]{new Direction(relativeEntityLoc.getDirection().normalize().multiply(meterLength))};
 			}
 		} else {
-			return new Direction[] {new Direction(horizontal ? Direction.IGNORE_PITCH : 0, yaw, ln)};
+			return new Direction[]{new Direction(isHorizontal ? Direction.IGNORE_PITCH : 0, yaw, meterLength)};
 		}
+		return new Direction[0];
 	}
-	
+
 	@Override
 	public boolean isSingle() {
 		return true;
 	}
-	
+
 	@Override
 	public Class<? extends Direction> getReturnType() {
 		return Direction.class;
 	}
-	
+
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		final Expression<?> relativeTo = this.relativeTo;
-		return (amount != null ? amount.toString(e, debug) + " meter(s) " : "") + (direction != null ? Direction.toString(direction) :
-				relativeTo != null ? " in " + (horizontal ? "horizontal " : "") + (facing ? "facing" : "direction") + " of " + relativeTo.toString(e, debug) :
-						(horizontal ? "horizontally " : "") + Direction.toString(0, yaw, 1));
+	public String toString(@Nullable Event event, boolean debug) {
+		return String.format("%s%s%s%s",
+			this.amount != null ? amount.toString(event, debug) + " meter(s) " : "",
+			(isHorizontal ? "horizontally " : ""), (isFacing ? "facing " : ""),
+			this.directionVector != null ? Direction.toString(directionVector) :
+				this.relativeTo != null ? "of " + relativeTo.toString(event, debug) :
+					Direction.toString(0, yaw, 1)
```

Comment:
I think it would look nicer if you assigned these variables first rather than having 90 ternaries in the return, it's not like we're saving lines doing this anyway.

Reply:
Will see what I can do